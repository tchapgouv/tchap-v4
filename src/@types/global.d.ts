/*
Copyright 2020-2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import "matrix-react-sdk/src/@types/global"; // load matrix-react-sdk's type extensions first
import type { Renderer } from "react-dom";
import type { logger } from "matrix-js-sdk/src/logger";

type ElectronChannel =
    | "app_onAction"
    | "before-quit"
    | "check_updates"
    | "install_update"
    | "ipcCall"
    | "ipcReply"
    | "loudNotification"
    | "preferences"
    | "seshat"
    | "seshatReply"
    | "setBadgeCount"
    | "update-downloaded"
    | "userDownloadCompleted"
    | "userDownloadAction"
    | "openDesktopCapturerSourcePicker"
    | "userAccessToken"
    | "serverSupportedVersions";

declare global {
    interface Window {
        mxSendRageshake: (text: string, withLogs?: boolean) => void;
        matrixLogger: typeof logger;
        matrixChat: ReturnType<Renderer>;

        // electron-only
        electron?: Electron;

        // opera-only
        opera?: any;

        // https://developer.mozilla.org/en-US/docs/Web/API/InstallTrigger
        InstallTrigger: any;
    }

    interface Electron {
        on(channel: ElectronChannel, listener: (event: Event, ...args: any[]) => void): void;
        send(channel: ElectronChannel, ...args: any[]): void;
    }
}

// add method which is missing from the node typing
declare module "url" {
    interface Url {
        format(): string;
    }
}
