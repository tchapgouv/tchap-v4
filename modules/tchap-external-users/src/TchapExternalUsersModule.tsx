/*
Copyright 2022 New Vector Ltd. t/a Element

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { RuntimeModule } from "@matrix-org/react-sdk-module-api/lib/RuntimeModule";
import { ModuleApi } from "@matrix-org/react-sdk-module-api/lib/ModuleApi";
import { CantAddDialog  } from "./components/CantAddDialog";
import { AddRoomListener, ExternalUsersLifecycle } from "@matrix-org/react-sdk-module-api/lib/tchap/lifecycles/ExternalUsers";
import { DialogProps } from "@matrix-org/react-sdk-module-api/lib/components/DialogContent";

import React from "react";

export default class TchapExternalUsersModule extends RuntimeModule {
    private domain: String;

    public constructor(moduleApi: ModuleApi) {
        super(moduleApi);

        // const rawConfig = this.moduleApi.getConfigValue(
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_NAMESPACE,
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_KEY,
        //   );
      
        //   assertValidGuestModuleConfig(rawConfig);
        //   this.config = rawConfig;

        this.on(ExternalUsersLifecycle.AddRoom, this.onAddRoom);
    }

    protected onAddRoom: AddRoomListener = async () => {
        await this.moduleApi.openDialog<null, DialogProps, CantAddDialog>("test", (props, ref) => <CantAddDialog ref={ref} {...props} />);
    };
}
