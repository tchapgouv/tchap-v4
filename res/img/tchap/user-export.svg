<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M19.5002 11.5094C19.8949 11.5094 20.281 11.4736 20.6555 11.4051C20.7505 11.9558 20.8 12.5221 20.8 13.1C20.8 18.5676 16.3676 23 10.9 23C5.43238 23 1 18.5676 1 13.1C1 7.63238 5.43238 3.2 10.9 3.2C11.7973 3.2 12.6666 3.31936 13.4931 3.54308C13.3177 4.11118 13.2234 4.71396 13.2234 5.3384C13.2234 8.74655 16.0336 11.5094 19.5002 11.5094ZM11.1308 21.4653C13.2927 21.4653 15.2551 20.6258 16.6993 19.2603C15.7939 17.0769 13.6417 15.5412 11.1308 15.5412C8.61988 15.5412 6.46768 17.0769 5.56225 19.2603C7.00643 20.6258 8.96888 21.4653 11.1308 21.4653ZM11.2799 7.81768C9.78854 7.95149 8.62035 9.18434 8.62035 10.6854C8.62035 12.2759 9.93179 13.5652 11.5495 13.5652C12.6276 13.5652 13.5696 12.9927 14.0781 12.1401C12.6655 11.1035 11.6462 9.57751 11.2799 7.81768Z" fill="currentColor"/>
<path d="M18.5033 0.999999C18.0923 0.999999 17.7502 1.34413 17.7502 1.74465L17.7502 5.79526L17.8105 7.01059L17.3115 6.41134L16.1704 5.1862C16.0409 5.04077 15.86 4.96364 15.6702 4.96364C15.306 4.96364 15 5.2367 15 5.61876C15 5.81535 15.0788 5.96079 15.2149 6.09806L17.9357 8.73968C18.1269 8.93251 18.3026 9 18.5033 9C18.6975 9 18.8716 8.93251 19.0709 8.73968L21.7918 6.09806C21.9264 5.96079 22 5.81535 22 5.61876C22 5.2367 21.7074 4.96364 21.3365 4.96364C21.1534 4.96364 20.9591 5.04077 20.8296 5.1862L19.6953 6.41134L19.1895 7.01059L19.2498 5.79526L19.2498 1.74465C19.2498 1.34413 18.9144 0.999999 18.5033 0.999999Z" fill="currentColor"/>
</svg>
