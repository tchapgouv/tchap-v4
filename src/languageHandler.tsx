/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import {
    IVariables,
    TranslatedString,
    TranslationKey as ReactTranslationKey,
    // eslint-disable-next-line camelcase
    _t as react_t,
    // eslint-disable-next-line camelcase
    _td as react_td,
    // eslint-disable-next-line camelcase
    _tDom as react_tDom,
    Tags,
    UserFriendlyError as ReactUserFriendlyError,
    ErrorOptions,
} from "matrix-react-sdk/src/languageHandler";
import { Leaves } from "matrix-react-sdk/src/@types/common";

import type ReactEN from "matrix-react-sdk/src/i18n/strings/en_EN.json";
import type EN from "./i18n/strings/en_EN.json";

/**
 * This module wraps languageHandler in the matrix-react-sdk and adds type casts to include translations
 * which we know will be injected by webpack.
 */

export type TranslationKey = Leaves<typeof EN & typeof ReactEN, "|", string | { other: string }, 4>;

export class UserFriendlyError extends ReactUserFriendlyError {
    public constructor(message: TranslationKey, substitutionVariablesAndCause?: IVariables & ErrorOptions) {
        super(message as ReactTranslationKey, substitutionVariablesAndCause);
    }
}

export function _td(s: TranslationKey): TranslationKey {
    return react_td(s as ReactTranslationKey);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _t(text: TranslationKey, variables?: IVariables): string;
export function _t(text: TranslationKey, variables: IVariables | undefined, tags: Tags): React.ReactNode;
export function _t(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_t(text as ReactTranslationKey, variables, tags!);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _tDom(text: TranslationKey, variables?: IVariables): TranslatedString;
export function _tDom(text: TranslationKey, variables: IVariables, tags: Tags): React.ReactNode;
export function _tDom(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_tDom(text as ReactTranslationKey, variables!, tags!);
}
