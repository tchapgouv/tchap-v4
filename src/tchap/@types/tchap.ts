export enum TchapRoomType {
    Direct = "direct",
    Private = "private",
    External = "external",
    Forum = "forum",
    Unknown = "unknown",
}

export enum TchapRoomAccessRule {
    Unrestricted = "unrestricted", // accessible to externals
    Restricted = "restricted", // not accessible to externals
}

export interface TchapIAccessRuleEventContent {
    rule: TchapRoomAccessRule; // eslint-disable-line camelcase
}

export const TchapRoomAccessRulesEventId = "im.vector.room.access_rules";
