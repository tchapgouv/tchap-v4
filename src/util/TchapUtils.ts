import { MatrixClientPeg } from "matrix-react-sdk/src/MatrixClientPeg";
import SdkConfig from "matrix-react-sdk/src/SdkConfig";
import AutoDiscoveryUtils from "matrix-react-sdk/src/utils/AutoDiscoveryUtils";
import { ValidatedServerConfig } from "matrix-react-sdk/src/utils/ValidatedServerConfig";

/**
 * Tchap utils.
 */

export default class TchapUtils {
    /**
         * Return a short value for getDomain().
         * @returns {string} The shortened value of getDomain().
         */
    static getShortDomain(): string {
        const cli = MatrixClientPeg.get();
        const baseDomain = cli.getDomain();
        const domain = baseDomain.split('.tchap.gouv.fr')[0].split('.').reverse().filter(Boolean)[0];

        return this.capitalize(domain) || 'Tchap';
    }

    /**
     * For the current user, get the room federation options.
     *
     * @returns { showRoomFederationOption: boolean, roomFederationDefault: boolean } options
     */
    static getRoomFederationOptions(): { showRoomFederationOption: boolean, roomFederationDefault: boolean } {
        const cli = MatrixClientPeg.get();
        const baseDomain = cli.getDomain();

        // Only show the federate switch to defense users : it's difficult to understand, so we avoid
        // displaying it unless it's really necessary.
        if (baseDomain === 'agent.intradef.tchap.gouv.fr') {
            return { showRoomFederationOption: true, roomFederationDefault: false };
        }

        return { showRoomFederationOption: false, roomFederationDefault: true };
    }

    /**
     * Capitalize a string.
     * @param {string} s The sting to capitalize.
     * @returns {string} The capitalized string.
     * @private
     */
    static capitalize(s: string): string {
        return s.charAt(0).toUpperCase() + s.slice(1);
    }

    static findHomeServerNameFromUrl = (url: string): string => {
        const homeServerList = SdkConfig.get()['homeserver_list'];
        const homeserver = homeServerList.find(homeServer => homeServer.base_url === url);
        return homeserver.server_name;
    };

    static randomHomeServer = () => {
        const homeServerList = SdkConfig.get()['homeserver_list'];
        return homeServerList[Math.floor(Math.random() * homeServerList.length)];
    };

    /**
     * Find the homeserver corresponding to the given email.
     * @param email Note : if email is invalid, this function still works and returns the externs server. (todo : fix)
     * @returns
     */
    static fetchHomeserverForEmail = async (email: string): Promise<void | {base_url: string, server_name: string}> => {
        const randomHomeServer = this.randomHomeServer();
        const infoUrl = "/_matrix/identity/api/v1/info?medium=email&address=";
        return fetch(randomHomeServer.base_url + infoUrl + email)
            .then((response) => {
                if (!response.ok) {
                    throw new Error('Could not find homeserver for this email');
                }
                return response.json();
            })
            .then(response => {
                // Never returns error : anything that doesn't match a homeserver (even invalid email) returns "externe".
                const serverUrl = "https://matrix." + response.hs;
                return {
                    base_url: serverUrl,
                    server_name: this.findHomeServerNameFromUrl(serverUrl),
                };
            })
            .catch((error) => {
                console.error('Could not find homeserver for this email', error);
                return;
            });
    };

    /**
     * Make a ValidatedServerConfig from the server urls.
     * Todo : merge this function with fetchHomeserverForEmail, they are always used together anyway.
     * @param
     * @returns
     */
    static makeValidatedServerConfig = (serverConfig): ValidatedServerConfig => {
        const discoveryResult = {
            "m.homeserver": {
                state: "SUCCESS",
                error: null,
                base_url: serverConfig.base_url,
                server_name: serverConfig.server_name,
            },
            "m.identity_server": {
                state: "SUCCESS",
                error: null,
                base_url: serverConfig.base_url, // On Tchap our Identity server urls and home server urls are the same
                server_name: serverConfig.server_name,
            },
        };
        const validatedServerConf = AutoDiscoveryUtils.buildValidatedConfigFromDiscovery(
            discoveryResult['m.homeserver'].server_name, discoveryResult);
        return validatedServerConf;
    };

    /**
     * Extract the name of the server from a home server address
     * @param serverName from "agent.dinum.beta.gouv.fr"
     * @returns to "Dinum"
     */
    static toFriendlyServerName = (serverName: string): string => {
        const serverUrl = "https://matrix." + serverName;
        const friendlyServerName = this.findHomeServerNameFromUrl(serverUrl);
        return this.capitalize(friendlyServerName);
    };

    /**
     * Ask the homeserver is cross signing is supported (async)
     * @returns Promise<true> is cross signing is supported by home server or false
     */
    static async isCrossSigningSupportedByServer(): Promise<boolean> {
        const cli = MatrixClientPeg.get();
        return cli.doesServerSupportUnstableFeature("org.matrix.e2e_cross_signing");
    }

    /**
     * @returns string The url to pass in next_link during registration. Compared to element-web, the hostname
     * is the homeserver instead of the tchap-web server. This changes the flow to avoid the redirection to
     * tchap-web, because tchap-web gets a "M_THREEPID_IN_USE" error from backend which is confusing.
     * We should fix this bug and remove this custom function.
     */
    static makeTchapRegistrationUrl(
        params: {client_secret: string, hs_url: string, is_url: string, session_id: string}): string {
        let url: string = params.hs_url + window.location.pathname + '#/register';

        const keys = Object.keys(params);
        for (let i = 0; i < keys.length; ++i) {
            if (i === 0) {
                url += '?';
            } else {
                url += '&';
            }
            const k = keys[i];
            url += k + '=' + encodeURIComponent(params[k]);
        }
        return url;
    }
}
