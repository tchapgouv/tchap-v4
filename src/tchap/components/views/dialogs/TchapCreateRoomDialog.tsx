/*
Copyright 2022 DINUM

Replaces CreateRoomDialog.tsx

Tchap has 3 types of rooms only : private, private open to externs, and public (forum).
Tchap does not support spaces for now, so there are no settings dependent on spaces (for example "Visible to space
members"). Element uses JoinRule.Restricted for this.
When/if Tchap supports spaces, we can decide on more specific settings if needed.

Note on imports : because this file will be copied to a different directory by the customisations
mechanism, imports must use absolute paths.
Except when importing from other customisation files. Then imports must use relative paths.
*/
import React, { ChangeEvent, createRef, KeyboardEvent } from "react";
import { Room } from "matrix-js-sdk/src/models/room";
import withValidation, { IFieldState } from "matrix-react-sdk/src/components/views/elements/Validation";
import { _t } from "matrix-react-sdk/src/languageHandler";
import { IOpts } from "matrix-react-sdk/src/createRoom";
import { getKeyBindingsManager } from "matrix-react-sdk/src/KeyBindingsManager";
import { KeyBindingAction } from "matrix-react-sdk/src/accessibility/KeyboardShortcuts";
import Field from "matrix-react-sdk/src/components/views/elements/Field";
import DialogButtons from "matrix-react-sdk/src/components/views/elements/DialogButtons";
import BaseDialog from "matrix-react-sdk/src/components/views/dialogs/BaseDialog";

import TchapUtils from "../../../util/TchapUtils";
import TchapRoomTypeSelector from "../elements/TchapRoomTypeSelector";
import { TchapRoomType } from "../../../@types/tchap";
import TchapCreateRoom from "../../../lib/createTchapRoom";

// We leave the same props as Element's version, to avoid unknown props warnings.
interface IProps {
    defaultPublic?: boolean; // unused for Tchap version
    defaultName?: string;
    parentSpace?: Room; 
    defaultEncrypted?: boolean; // unused for Tchap version
    onFinished(proceed: boolean, opts?: IOpts): void;
}

interface IState {
    name: string;
    nameIsValid: boolean;
    tchapRoomType: TchapRoomType;
    forumFederationSwitchValue: boolean;
    showFederateSwitch: boolean;
    createRoomInSpace: boolean;
}

export default class TchapCreateRoomDialog extends React.Component<IProps, IState> {
    private nameField = createRef<Field>();
    private readonly createRoomInSpace: boolean;

    public constructor(props) {
        super(props);
        this.createRoomInSpace = !!this.props.parentSpace;

        const federationOptions = TchapUtils.getRoomFederationOptions();

        this.state = {
            name: this.props.defaultName || "",
            nameIsValid: false,
            tchapRoomType: TchapRoomType.Private,
            forumFederationSwitchValue: federationOptions.forumFederationSwitchDefaultValue,
            showFederateSwitch: federationOptions.showForumFederationSwitch,
            createRoomInSpace: createRoomInSpace,
        };
    }

    componentDidMount() {
        // move focus to first field when showing dialog
        this.nameField.current.focus();
    }

    private onCancel = () => {
        this.props.onFinished(false);
    };

    private onForumFederatedChange = (forumFederationSwitchValue: boolean): void => {
        this.setState({ forumFederationSwitchValue });
    };

    private onTchapRoomTypeChange = (tchapRoomType: TchapRoomType): void => {
        this.setState({ tchapRoomType: tchapRoomType });
    };

    private onNameChange = (ev: ChangeEvent<HTMLInputElement>): void => {
        this.setState({ name: ev.target.value });
    };

    private onNameValidate = async (fieldState: IFieldState) => {
        const result = await TchapCreateRoomDialog.validateRoomName(fieldState);
        this.setState({ nameIsValid: result.valid });
        return result;
    };

    private static validateRoomName = withValidation({
        rules: [
            {
                key: "required",
                test: async ({ value }) => !!value,
                invalid: () => _t("Please enter a name for the room"),
            },
        ],
    });

    private onKeyDown = (event: KeyboardEvent): void => {
        const action = getKeyBindingsManager().getAccessibilityAction(event);
        switch (action) {
            case KeyBindingAction.Enter:
                this.onOk();
                event.preventDefault();
                event.stopPropagation();
                break;
        }
    };

    private isSelectedRoomFederated = (): boolean => {
        return this.state.tchapRoomType === TchapRoomType.Forum && this.state.showFederateSwitch
            ? this.state.forumFederationSwitchValue
            : true;
    };

    private onOk = async () => {
        const activeElement = document.activeElement as HTMLElement;
        if (activeElement) {
            activeElement.blur();
        }
        await this.nameField.current.validate({ allowEmpty: false });
        // Validation and state updates are async, so we need to wait for them to complete
        // first. Queue a `setState` callback and wait for it to resolve.
        await new Promise<void>((resolve) => this.setState({}, resolve));
        if (this.state.nameIsValid) {
            this.props.onFinished(
                true,
                TchapCreateRoom.roomCreateOptions(
                    this.state.name,
                    this.state.tchapRoomType,
                    this.isSelectedRoomFederated(),
                    this.props.parentSpace
                ),
            );
        } else {
            let field;
            if (!this.state.nameIsValid) {
                field = this.nameField.current;
            }
            if (field) {
                field.focus();
                field.validate({ allowEmpty: false, focused: true });
            }
        }
    };

    public render() {
        const shortDomain: string = TchapUtils.getShortDomain();

        const title = this.createRoomInSpace ? _t("Create a room in this space") : _t("Create a room");

        return (
            <BaseDialog
                className="tc_TchapCreateRoomDialog"
                onFinished={this.props.onFinished}
                title={title}
                screenName="CreateRoom"
            >
                <form onSubmit={this.onOk} onKeyDown={this.onKeyDown}>
                    <div className="mx_Dialog_content">
                        <Field
                            ref={this.nameField}
                            label={_t("Name")}
                            onChange={this.onNameChange}
                            onValidate={this.onNameValidate}
                            value={this.state.name}
                        />

                        <TchapRoomTypeSelector
                            value={this.state.tchapRoomType}
                            label={_t("Type of room")}
                            showFederateSwitch={this.state.showFederateSwitch}
                            shortDomain={shortDomain}
                            forumFederationSwitchValue={this.state.forumFederationSwitchValue}
                            setForumFederationSwitchValue={this.onForumFederatedChange}
                            setRoomType={this.onTchapRoomTypeChange}
                            createRoomInSpace={this.createRoomInSpace}
                        />
                    </div>
                </form>
                <DialogButtons
                    primaryButton={_t("Create Room")}
                    onPrimaryButtonClick={this.onOk}
                    onCancel={this.onCancel}
                />
            </BaseDialog>
        );
    }
}
