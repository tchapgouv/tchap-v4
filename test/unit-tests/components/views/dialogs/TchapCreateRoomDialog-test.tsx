import React from "react";
// eslint-disable-next-line deprecate/import
import { mount, ReactWrapper } from "enzyme";
// eslint-disable-next-line deprecate/import
import { act } from "react-dom/test-utils";
import toJson from "enzyme-to-json";
import { MatrixClientPeg } from "matrix-react-sdk/src/MatrixClientPeg";
import { EventEmitter } from "events";

import { TchapRoomType } from "../../../../../src/tchap/@types/tchap";
import TchapUtils from "../../../../../src/tchap/util/TchapUtils";
import TchapCreateRoomDialog from "../../../../../src/components/views/dialogs/TchapCreateRoomDialog";

//mocking module with jest.mock should be done outside the test. Before any import of the mocked module.
//I could not make a mock of TchapCreateRoomDialog, the real implementation was taken each time. Then I used jest spyOn

/*    return {
    jest.mock('../../src/lib/createTchapRoom',() => {
        __esModule: true,
        default: jest.fn(() => 42),
        roomCreateOptions: jest.fn(() => 43),
    };
});
*/

/*
 * This unit test targets the TchapCreateRoomDialog react component. It is mounted virtually to be tested thanks to enzyme framework.
 * With the act function, a state is set inside the component, meanwhile the submitForm function submits the form in order to make action on the internal state of the component.
 * No html snapshot is used in this test. Jest mocking is used to simulate behaviour of depending modules such as TchapUtils, TchapCreateRoomDialog, MatrixClientPeg
 */
describe("TchapCreateRoomDialog", () => {
    let mockClient;
    const roomName = "roomName";

    // default props for the component, will be overriden
    const defaultProps = {
        defaultPublic: undefined, // unused for Tchap version
        defaultName: undefined, // unused for Tchap version
        parentSpace: undefined, // unused for Tchap version
        defaultEncrypted: undefined, // unused for Tchap version
        onFinished: jest.fn(),
    };

    //simulate the submit of the form
    const submitForm = async (wrapper: ReactWrapper) => {
        act(() => {
            wrapper.find("form").simulate("submit", { preventDefault: () => {} });
        });
        await new Promise(process.nextTick);
    };

    // build a new component using enzyme
    const getComponent = (props = {}): ReactWrapper => mount(<TchapCreateRoomDialog {...defaultProps} {...props} />);

    /*
    // Note : you can also build a shallow component https://fr.reactjs.org/docs/shallow-renderer.html
    // can be used for simple component
    // not used in this test
    const getShallowComponent = (props={}) => {
        shallow(<TchapCreateRoomDialog {...defaultProps} {...props} />);
    }
    */

    beforeEach(() => {
        jest.resetAllMocks();

        //mock matrix client
        mockClient = new MockClient();
        jest.spyOn(MatrixClientPeg, "get").mockReturnValue(mockClient);

        //mock tchap utils
        jest.spyOn(TchapUtils, "getShortDomain").mockReturnValue("AGENT");
        jest.spyOn(TchapUtils, "getRoomFederationOptions").mockReturnValue({
            showForumFederationSwitch: true,
            forumFederationSwitchDefaultValue: false,
        });
    });

    it("should render the whole component with with the allow access switch", () => {
        jest.spyOn(TchapUtils, "getRoomFederationOptions").mockReturnValue({
            showForumFederationSwitch: true,
            forumFederationSwitchDefaultValue: false,
        });
        const component = getComponent();
        const allowAccessSwitch = component.find(".mx_SettingsFlag");
        expect(toJson(allowAccessSwitch)).toMatchSnapshot(
            "allow access switch is present, allows other domain than AGENT",
        );
    });

    it("should render the room dialog without the allow access switch", () => {
        jest.spyOn(TchapUtils, "getRoomFederationOptions").mockReturnValue({
            showForumFederationSwitch: false,
            forumFederationSwitchDefaultValue: false,
        });
        const component = getComponent();
        const allowAccessSwitch = component.find(".mx_SettingsFlag");
        expect(allowAccessSwitch).toEqual({});
    });

    it("Should not create any room wihout a name", async () => {
        const onFinished = jest.fn();
        const wrapper = getComponent({ onFinished });
        // set state in component
        act(() => {
            wrapper.setState({
                name: "",
                tchapRoomType: TchapRoomType.Private,
            });
        });

        await submitForm(wrapper);

        expect(onFinished).toBeCalledTimes(0);
    });

    it("Should create a room with default value", async () => {
        const onFinished = jest.fn();
        const defaultName = "defaultName";
        const wrapper = getComponent({ onFinished, defaultName });
        // set state in component
        act(() => {
            wrapper.setState({
                tchapRoomType: TchapRoomType.Private,
            });
        });

        await submitForm(wrapper);

        expect(onFinished).toBeCalledTimes(1);
    });

    it("Should create a private room", async () => {
        const onFinished = jest.fn();

        const privateRoomExpectedOpts = {
            createOpts: {
                name: roomName,
                creation_content: {
                    "m.federate": true,
                },
                initial_state: [
                    {
                        content: {
                            rule: "restricted",
                        },
                        state_key: "",
                        type: "im.vector.room.access_rules",
                    },
                ],
                visibility: "private",
                preset: "private_chat",
            },
            guestAccess: false,
            joinRule: "invite",
            encryption: true,
            historyVisibility: "invited",
        };

        const wrapper = getComponent({ onFinished });

        // set state in component
        act(() => {
            wrapper.setState({
                name: roomName,
                tchapRoomType: TchapRoomType.Private,
                showFederateSwitch: false,
            });
        });

        await submitForm(wrapper);

        expect(onFinished).toHaveBeenCalledWith(true, privateRoomExpectedOpts);
    });

    it("Should create a public room without federation", async () => {
        const onFinished = jest.fn();

        const publicRoomWithoutFederationExpectedOpts = {
            createOpts: {
                name: roomName,
                creation_content: {
                    "m.federate": false,
                },
                initial_state: [
                    {
                        content: {
                            rule: "restricted",
                        },
                        state_key: "",
                        type: "im.vector.room.access_rules",
                    },
                ],
                visibility: "public",
                preset: "public_chat",
            },
            guestAccess: false,
            joinRule: "public",
            encryption: false,
            historyVisibility: "shared",
        };
        const wrapper = getComponent({ onFinished });

        // set state in component
        act(() => {
            wrapper.setState({
                name: roomName,
                tchapRoomType: TchapRoomType.Forum,
                forumFederationSwitchValue: false,
                showFederateSwitch: true,
            });
        });

        await submitForm(wrapper);

        expect(onFinished).toHaveBeenCalledWith(true, publicRoomWithoutFederationExpectedOpts);
    });

    it("Should create a public room with federation and switch", async () => {
        const onFinished = jest.fn();

        const publicRoomWithFederationExpectedOpts = {
            createOpts: {
                name: roomName,
                creation_content: {
                    "m.federate": true,
                },
                initial_state: [
                    {
                        content: {
                            rule: "restricted",
                        },
                        state_key: "",
                        type: "im.vector.room.access_rules",
                    },
                ],
                visibility: "public",
                preset: "public_chat",
            },
            guestAccess: false,
            joinRule: "public",
            encryption: false,
            historyVisibility: "shared",
        };
        const wrapper = getComponent({ onFinished });

        // set state in component
        act(() => {
            wrapper.setState({
                name: roomName,
                tchapRoomType: TchapRoomType.Forum,
                forumFederationSwitchValue: true,
                showFederateSwitch: true,
            });
        });

        await submitForm(wrapper);

        expect(onFinished).toHaveBeenCalledWith(true, publicRoomWithFederationExpectedOpts);
    });

    it("Should create a public room with federation but no switch", async () => {
        const onFinished = jest.fn();

        const publicRoomWithFederationExpectedOpts = {
            createOpts: {
                name: roomName,
                creation_content: {
                    "m.federate": true,
                },
                initial_state: [
                    {
                        content: {
                            rule: "restricted",
                        },
                        state_key: "",
                        type: "im.vector.room.access_rules",
                    },
                ],
                visibility: "public",
                preset: "public_chat",
            },
            guestAccess: false,
            joinRule: "public",
            encryption: false,
            historyVisibility: "shared",
        };
        const wrapper = getComponent({ onFinished });

        // set state in component
        act(() => {
            wrapper.setState({
                name: roomName,
                tchapRoomType: TchapRoomType.Forum,
                showFederateSwitch: false,
            });
        });

        await submitForm(wrapper);

        expect(onFinished).toHaveBeenCalledWith(true, publicRoomWithFederationExpectedOpts);
    });

    it("Should create an external room", async () => {
        const onFinished = jest.fn();

        const externalRoomExpectedOpts = {
            createOpts: {
                name: roomName,
                creation_content: {
                    "m.federate": true,
                },
                initial_state: [
                    {
                        content: {
                            rule: "unrestricted",
                        },
                        state_key: "",
                        type: "im.vector.room.access_rules",
                    },
                ],
                visibility: "private",
                preset: "private_chat",
            },
            guestAccess: false,
            joinRule: "invite",
            encryption: true,
            historyVisibility: "invited",
        };
        const wrapper = getComponent({ onFinished });

        // set state in component
        act(() => {
            wrapper.setState({
                name: roomName,
                tchapRoomType: TchapRoomType.External,
                showFederateSwitch: false,
            });
        });

        await submitForm(wrapper);

        expect(onFinished).toHaveBeenCalledWith(true, externalRoomExpectedOpts);
    });
});

class MockClient extends EventEmitter {
    // getUserId = jest.fn();
    // getKeyBackupVersion = jest.fn().mockResolvedValue(undefined);
    // getRooms = jest.fn().mockReturnValue([]);
    // doesServerSupportUnstableFeature = jest.fn().mockResolvedValue(true);
    // isCrossSigningReady = jest.fn().mockResolvedValue(true);
    // isSecretStorageReady = jest.fn().mockResolvedValue(true);
    // isCryptoEnabled = jest.fn().mockReturnValue(true);
    // isInitialSyncComplete = jest.fn().mockReturnValue(true);
    // getKeyBackupEnabled = jest.fn();
    // getStoredDevicesForUser = jest.fn().mockReturnValue([]);
    // getCrossSigningId = jest.fn();
    // getStoredCrossSigningForUser = jest.fn();
    // waitForClientWellKnown = jest.fn();
    // downloadKeys = jest.fn();
    // isRoomEncrypted = jest.fn();
    // getClientWellKnown = jest.fn();
    getDomain = jest.fn();
}
