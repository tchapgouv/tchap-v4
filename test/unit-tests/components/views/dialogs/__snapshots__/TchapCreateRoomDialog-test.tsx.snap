// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TchapCreateRoomDialog should render the whole component with with the allow access switch: allow access switch is present, allows other domain than AGENT 1`] = `
<div
  className="mx_SettingsFlag"
>
  <span
    className="mx_SettingsFlag_label"
  >
    Allow access to this room to all users, even outside "AGENT" domain
  </span>
  <_default
    checked={false}
    onChange={[Function]}
    title="Allow access to this room to all users, even outside "AGENT" domain"
  >
    <AccessibleTooltipButton
      aria-checked={false}
      aria-disabled={false}
      className="mx_ToggleSwitch mx_ToggleSwitch_enabled"
      onClick={[Function]}
      role="switch"
      title="Allow access to this room to all users, even outside "AGENT" domain"
    >
      <AccessibleButton
        aria-checked={false}
        aria-disabled={false}
        aria-label="Allow access to this room to all users, even outside "AGENT" domain"
        className="mx_ToggleSwitch mx_ToggleSwitch_enabled"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        role="switch"
        tabIndex={0}
      >
        <div
          aria-checked={false}
          aria-disabled={false}
          aria-label="Allow access to this room to all users, even outside "AGENT" domain"
          className="mx_AccessibleButton mx_ToggleSwitch mx_ToggleSwitch_enabled"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          role="switch"
          tabIndex={0}
        >
          <div
            className="mx_ToggleSwitch_ball"
          />
        </div>
      </AccessibleButton>
    </AccessibleTooltipButton>
  </_default>
</div>
`;
